File: sim/src/core/city/tensor.ts
/**
 * Tensor field for street orientation — Chen et al., "Interactive Procedural
 * Street Modeling" (SIGGRAPH 2008). A 2D symmetric traceless tensor is stored
 * in angle-doubled form (cos 2θ, sin 2θ) so that θ and θ+π blend as the same
 * street direction. Basis fields (grid patches, one radial center, water
 * boundaries) are blended with distance falloffs; streets trace the major
 * eigenvector (θ) or the minor one (θ + π/2).
 */
import type { Vec2 } from '../geometry';

export interface GridBasis {
  center: Vec2;
  theta: number;
  /** Gaussian falloff radius, meters */
  sigma: number;
  weight: number;
}

export interface BoundarySample {
  pos: Vec2;
  /** tangent angle of the shoreline at this sample */
  theta: number;
}

export interface TensorField {
  grids: GridBasis[];
  /** citywide constant orientation (no falloff) — makes rigid grid cities read
   *  as one coherent grid instead of patches the radial field bends apart */
  globalGrid?: { theta: number; weight: number };
  radialCenter: Vec2;
  radialWeight: number;
  radialSigma: number;
  boundaries: BoundarySample[];
  boundarySigma: number;
  boundaryWeight: number;
  /** small angular noise to break perfect regularity */
  noise: (x: number, y: number) => number;
  noiseWeight: number;
}

/** Street direction (major eigenvector angle) at a point. */
export function sampleAngle(f: TensorField, p: Vec2): number {
  let c2 = 0;
  let s2 = 0;

  if (f.globalGrid) {
    c2 += f.globalGrid.weight * Math.cos(2 * f.globalGrid.theta);
    s2 += f.globalGrid.weight * Math.sin(2 * f.globalGrid.theta);
  }

  for (const g of f.grids) {
    const dx = p.x - g.center.x;
    const dy = p.y - g.center.y;
    const w = g.weight * Math.exp(-(dx * dx + dy * dy) / (2 * g.sigma * g.sigma));
    c2 += w * Math.cos(2 * g.theta);
    s2 += w * Math.sin(2 * g.theta);
  }

  {
    const dx = p.x - f.radialCenter.x;
    const dy = p.y - f.radialCenter.y;
    const d = Math.sqrt(dx * dx + dy * dy) + 1;
    const theta = Math.atan2(dy, dx); // radial: streets point at the center
    const w = f.radialWeight * Math.exp(-d / f.radialSigma);
    c2 += w * Math.cos(2 * theta);
    s2 += w * Math.sin(2 * theta);
  }

  for (const b of f.boundaries) {
    const dx = p.x - b.pos.x;
    const dy = p.y - b.pos.y;
    const d = Math.sqrt(dx * dx + dy * dy);
    if (d > f.boundarySigma * 3) continue;
    const w = f.boundaryWeight * Math.exp(-d / f.boundarySigma);
    c2 += w * Math.cos(2 * b.theta);
    s2 += w * Math.sin(2 * b.theta);
  }

  let theta = 0.5 * Math.atan2(s2, c2);
  theta += (f.noise(p.x, p.y) - 0.5) * f.noiseWeight;
  return theta;
}

/**
 * Direction to step in at p, following the field along `eigen` (0 = major,
 * 1 = minor/perpendicular), keeping continuity with the previous direction
 * (eigenvectors are sign-ambiguous).
 */
export function stepDirection(f: TensorField, p: Vec2, eigen: 0 | 1, prev: Vec2 | null): Vec2 {
  let theta = sampleAngle(f, p);
  if (eigen === 1) theta += Math.PI / 2;
  let dx = Math.cos(theta);
  let dy = Math.sin(theta);
  if (prev && dx * prev.x + dy * prev.y < 0) {
    dx = -dx;
    dy = -dy;
  }
  return { x: dx, y: dy };
}

File: crates/mf-sim/src/city/tensor.rs
//! Tensor field for street orientation. Port of `sim/src/core/city/tensor.ts`.
//!
//! Chen et al., "Interactive Procedural Street Modeling" (SIGGRAPH 2008). A 2D
//! symmetric traceless tensor is stored in angle-doubled form so that theta and
//! theta+pi blend as the same street direction. Basis fields (grid patches, one
//! radial center, water boundaries) blend with distance falloffs; streets trace
//! the major eigenvector (theta) or the minor one (theta + pi/2).

use crate::geometry::Vec2;

/// A grid-orientation basis patch. Mirrors `GridBasis`.
#[derive(Clone, Copy, Debug)]
pub struct GridBasis {
    /// Patch center.
    pub center: Vec2,
    /// Orientation angle (radians).
    pub theta: f64,
    /// Gaussian falloff radius, meters.
    pub sigma: f64,
    /// Blend weight.
    pub weight: f64,
}

/// A shoreline tangent sample. Mirrors `BoundarySample`.
#[derive(Clone, Copy, Debug)]
pub struct BoundarySample {
    /// Sample position.
    pub pos: Vec2,
    /// Tangent angle of the shoreline at this sample.
    pub theta: f64,
}

/// The blended tensor field. Mirrors `TensorField`. The angular `noise` source
/// is a boxed closure (the TS field is a function); it borrows the detail
/// noise, so the field carries a lifetime.
pub struct TensorField<'a> {
    /// Grid patches.
    pub grids: Vec<GridBasis>,
    /// Citywide constant orientation (theta, weight); no falloff.
    pub global_grid: Option<(f64, f64)>,
    /// Radial convergence center.
    pub radial_center: Vec2,
    /// Radial blend weight.
    pub radial_weight: f64,
    /// Radial falloff length, meters.
    pub radial_sigma: f64,
    /// Shoreline boundary samples.
    pub boundaries: Vec<BoundarySample>,
    /// Boundary falloff length, meters.
    pub boundary_sigma: f64,
    /// Boundary blend weight.
    pub boundary_weight: f64,
    /// Small angular noise source.
    pub noise: Box<dyn Fn(f64, f64) -> f64 + 'a>,
    /// Noise weight.
    pub noise_weight: f64,
}

/// Street direction (major eigenvector angle) at a point. Mirrors `sampleAngle`.
pub fn sample_angle(f: &TensorField, p: Vec2) -> f64 {
    let mut c2 = 0.0;
    let mut s2 = 0.0;

    if let Some((theta, weight)) = f.global_grid {
        c2 += weight * (2.0 * theta).cos();
        s2 += weight * (2.0 * theta).sin();
    }

    for g in &f.grids {
        let dx = p.x - g.center.x;
        let dy = p.y - g.center.y;
        let w = g.weight * (-(dx * dx + dy * dy) / (2.0 * g.sigma * g.sigma)).exp();
        c2 += w * (2.0 * g.theta).cos();
        s2 += w * (2.0 * g.theta).sin();
    }

    {
        let dx = p.x - f.radial_center.x;
        let dy = p.y - f.radial_center.y;
        let d = (dx * dx + dy * dy).sqrt() + 1.0;
        let theta = dy.atan2(dx); // radial: streets point at the center
        let w = f.radial_weight * (-d / f.radial_sigma).exp();
        c2 += w * (2.0 * theta).cos();
        s2 += w * (2.0 * theta).sin();
    }

    for b in &f.boundaries {
        let dx = p.x - b.pos.x;
        let dy = p.y - b.pos.y;
        let d = (dx * dx + dy * dy).sqrt();
        if d > f.boundary_sigma * 3.0 {
            continue;
        }
        let w = f.boundary_weight * (-d / f.boundary_sigma).exp();
        c2 += w * (2.0 * b.theta).cos();
        s2 += w * (2.0 * b.theta).sin();
    }

    let mut theta = 0.5 * s2.atan2(c2);
    theta += ((f.noise)(p.x, p.y) - 0.5) * f.noise_weight;
    theta
}

/// Direction to step in at `p`, following `eigen` (0 = major, 1 = minor),
/// keeping continuity with `prev`. Mirrors `stepDirection`.
pub fn step_direction(f: &TensorField, p: Vec2, eigen: u8, prev: Option<Vec2>) -> Vec2 {
    let mut theta = sample_angle(f, p);
    if eigen == 1 {
        theta += std::f64::consts::FRAC_PI_2;
    }
    let mut dx = theta.cos();
    let mut dy = theta.sin();
    if let Some(pv) = prev {
        if dx * pv.x + dy * pv.y < 0.0 {
            dx = -dx;
            dy = -dy;
        }
    }
    Vec2 { x: dx, y: dy }
}
